From 65e47fd58d933b75a2e75777eb36e28873d98be3 Mon Sep 17 00:00:00 2001
From: isabel <isabel@isabelroses.com>
Date: Thu, 5 Dec 2024 17:05:38 +0000
Subject: [PATCH] feat(cli): add callPackage flag

adds a --call-package or -C cli option to build a package from the cli
based on the work of https://github.com/privatevoid-net/nix-super
---
 lix/libcmd/command.hh      |  1 +
 lix/libcmd/installables.cc | 25 +++++++++++++++++++++----
 2 files changed, 22 insertions(+), 4 deletions(-)

diff --git a/lix/libcmd/command.hh b/lix/libcmd/command.hh
index 478032116..40ee997db 100644
--- a/lix/libcmd/command.hh
+++ b/lix/libcmd/command.hh
@@ -111,6 +111,7 @@ struct SourceExprCommand : virtual Args, MixFlakeOptions
 {
     std::optional<Path> file;
     std::optional<std::string> expr;
+    std::optional<Path> callPackageFile;
 
     SourceExprCommand();
 
diff --git a/lix/libcmd/installables.cc b/lix/libcmd/installables.cc
index 700f5115a..8141d6ab4 100644
--- a/lix/libcmd/installables.cc
+++ b/lix/libcmd/installables.cc
@@ -165,6 +165,19 @@ SourceExprCommand::SourceExprCommand()
         .labels = {"expr"},
         .handler = {&expr}
     });
+
+    addFlag({
+        .longName = "call-package",
+        .shortName = 'C',
+        .description =
+            "Interpret [*installables*](@docroot@/command-ref/new-cli/nix.md#installables) as attribute paths relative to the callPackageable Nix expression stored in *file*. "
+            "The `callPackage` function is taken from `<nixpkgs>`. "
+            "Implies `--impure`.",
+        .category = installablesCategory,
+        .labels = {"file"},
+        .handler = {&callPackageFile},
+        .completer = completePath
+    });
 }
 
 MixReadOnlyOption::MixReadOnlyOption()
@@ -428,12 +441,12 @@ Installables SourceExprCommand::parseInstallables(
 {
     Installables result;
 
-    if (file || expr) {
-        if (file && expr)
-            throw UsageError("'--file' and '--expr' are exclusive");
+    if (file || expr || callPackageFile) {
+        if ((file && expr) || (file && callPackageFile) || (expr && callPackageFile))
+            throw UsageError("'--file', '--expr' and '--call-package' are exclusive");
 
         // FIXME: backward compatibility hack
-        if (file) {
+        if (file || callPackageFile) {
             evalSettings.pureEval.override(false);
             getEvalState()->paths.allowedPaths.reset();
         }
@@ -447,6 +460,10 @@ Installables SourceExprCommand::parseInstallables(
         }
         else if (file)
             state.evalFile(lookupFileArg(state, *file), *vFile);
+        else if (callPackageFile) {
+            auto & e = state.parseExprFromString(fmt("(import <nixpkgs> {}).callPackage %s {}", CanonPath::fromCwd(*callPackageFile)), CanonPath::fromCwd());
+            state.eval(e, *vFile);
+        }
         else {
             auto & e = state.parseExprFromString(*expr, CanonPath::fromCwd());
             state.eval(e, *vFile);
-- 
2.44.1

